modules/issue_tracker/php/uploadhelper.class.inc

```diff
@@ -0,0 +1,209 @@
+                'tmp_name'    => $files['file']['tmp_name'],
+                'size'        => $this->_bytes,
+                'inserted_by' => $this->_user->getData('UserID'),
+                'description' => $this->_values['fileDescription'] ?? '',
+            );
+
+        $this->setFullPath($this->_fileToUpload);
+
+        $this->begin();
+        $this->registerFile($this->_fileToUpload);
+        $this->endWithSuccess();
+
+        return array('message' => 'success');
+    }
+
+    /**
+     * Sets $fileToUpload->full_path
+     *
+     * @param object $fileToUpload The file to upload
+     *
+     * @return void
+     * @throws \LorisException
+     */
+    function setFullPath(Object &$fileToUpload) : void
+    {
+        $config = \NDB_Config::singleton();
+        $attachment_data_dir = rtrim(
+            $config->getSetting('IssueTrackerDataPath'),
+            '/'
+        );
+
+        if (is_writable($attachment_data_dir)) {
+            $fileToUpload->full_path = $attachment_data_dir .
+                                       '/attachments/' .
+                                       strval($this->_values['issueID']);
+            if (!is_dir($fileToUpload->full_path)) {
+                mkdir($fileToUpload->full_path, 0770, true);
+            }
+        } else {
+            throw new \LorisException(
+                'Issue_Tracker attachments directory not writable.'
+            );
+        }
+
+    }
+
+    /**
+     * This insert a record in the issues_file_collection table
+     *
+     * @param object $fileToUpload The object containing the $_FILES
+     *                             and the $_POST values.
+     *
+     * @return void
+     */
+    function registerFile(Object &$fileToUpload) : void
+    {
+        $DB     =& \Database::singleton();
+        $values = array(
+            'issueID'     => $this->_values['issueID'],
+            'file_uuid'   => $fileToUpload->uuid,
+            'date_added'  => date('Y-m-d h:i:s', time()),
+            'file_name'   => $fileToUpload->file,
+            'deleted'     => 0,
+            'user'        => $fileToUpload->inserted_by,
+            'description' => $fileToUpload->description,
+            'file_size'   => $fileToUpload->size,
+        );
+        try {
+            $DB->insert('issues_attachments', $values);
+            // Move file from tmp to issue_tracker attachment directory.
+            $path = $fileToUpload->full_path . '/' . $fileToUpload->uuid;
+            move_uploaded_file($_FILES['file']['tmp_name'], $path);
+        } catch (\DatabaseException $e) {
+            $this->endWithFailure();
+            error_log("DatabaseException: $e");
+            $this->_response = json_encode(
+                array(
+                    'message' => 'File registration failed.
+                    Is the description empty?',
+                    'error'   => true,
+                )
+            );
+        }
+    }
+
+    /**
+     * Create a UUID v4 string.
```

Comment:
Version 4 (random) isn't very appropriate for this use case. Version 3 or 5 (content hash based) would be better. If the same data is uploaded multiple times, there's no reason to duplicate it on disk under multiple different names. On that note, I don't see why it needs to be a UUID. Any content hash should do since the content isn't user-facing (and unfortunately both uuid v3 and v5 are based on outdated hashing algorithms.)

Reply:
@driusan I believe this was discussed at the LORIS meeting. The reason for using UUD v4 & not using just any random hash algorithm that generates a hash from the filename or content is for the general use case of multiple attachments being added to issues. Multiple different users can be adding attachments to different issues and where the file can be the same.

Reply:
that reasoning doesn't make any sense. If multiple different users upload the same attachment there's no reason to duplicate the file content on the filesystem, the content is the same and it's served to the client under the filename from the attachments table.

Reply:
@driusan so you think when a user uploads an image on lets say googledrive and another user uploads the same image. That it makes more sense for that image to just be stored in one place? I mean to me that doesn't make sense even with it saving space.

Reply:
Yes it makes more sense for that image to just be stored in one place. Duplicating it just because different people uploaded the same thing makes no sense.